# Práctica Clase 1 

1) Crear una función capaz de convertir números enteros de base 10 a base 2. Debe recibir como parámetro el número a convertir<br>
Consideraciones:<br> 
a. Tratar de resolverlo sin usar la función format(nro,"b")<br>
b. El pdf "conversion-de-decimal-a-binario.pdf" puede servir de ayuda.


In [17]:
def NumeroBinario(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    if type(numero) != int:
        return None
     
    if numero < 0:
        return None

    elif numero == 0:
        return 0

    lista = []                   # Almacenamiento de los restos
    div_resto = 0

    while (numero > 0):          # Mientras que el numero sea mayor que 0, vamos a hacer la division sucesiva
        div_resto = numero % 2   # Sacamos el resto de dividir por dos
        numero = numero // 2     # Se divide sucesivamente entre dos
        lista.append(div_resto)  # Guardamos el resto conseguido

    numero_binario = ''           # Esto vamos a retornar
    for elemento in lista:
        numero_binario = str(elemento) +  numero_binario 
    
    # lista = lista[::-1]

    return int(numero_binario)


In [11]:
#Ejemplo:
NumeroBinario(29)

11101

In [12]:
NumeroBinario(1)

1

In [13]:
NumeroBinario(1000)

1111101000

In [18]:
NumeroBinario('veintidos') #22

In [19]:
format(29, 'b')

'11101'

In [20]:
bin(29)

'0b11101'

2) Convertir de decimal a binario las fracciones 1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9. Luego analizar los resultados y observar qué particularidad se encuentra en los mismos. Se puede usar Python o una calculadora, lo importante es ver si hay algo que podemos notar...

In [30]:
float(0.625 * 2) - 1

0.25

In [31]:
def NumeroBinarioFraccionario(decimal):
    if (decimal >= 1) or (decimal < 0):
        return None
    elif decimal == 0:
        return 0
    
    lim_periodico = 24               # Número máximo de decimales
    lista = []

    while decimal > 0 and lim_periodico > 0:
        lista.append(int(decimal * 2))
        decimal = decimal * 2 - (int(decimal * 2))
        lim_periodico -= 1

    numero_binario = '0.'
    for elemento in lista:
        numero_binario = numero_binario + str(elemento)

    return numero_binario

In [34]:
for i in range(2,10):
    print(f'Fracion 1/{i} : {1/i} y en Binario: {NumeroBinarioFraccionario(1/i)}')

Fracion 1/2 : 0.5 y en Binario: 0.1
Fracion 1/3 : 0.3333333333333333 y en Binario: 0.010101010101010101010101
Fracion 1/4 : 0.25 y en Binario: 0.01
Fracion 1/5 : 0.2 y en Binario: 0.001100110011001100110011
Fracion 1/6 : 0.16666666666666666 y en Binario: 0.001010101010101010101010
Fracion 1/7 : 0.14285714285714285 y en Binario: 0.001001001001001001001001
Fracion 1/8 : 0.125 y en Binario: 0.001
Fracion 1/9 : 0.1111111111111111 y en Binario: 0.000111000111000111000111


In [39]:
import numpy as np
NumeroBinarioFraccionario(1/17)

'0.000011110000111100001111'

Salida esperada:
* Fraccion 1 / 2 :  0.5  y En binario:  0.1
* Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
* Fraccion 1 / 4 :  0.25  y En binario:  0.01
* Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
* Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
* Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
* Fraccion 1 / 8 :  0.125  y En binario:  0.001
* Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111

In [ ]:
"""
No Data
N/D
ND
-------
Null
None
"""

"""
Nombre: Gonzalo 
Apellido: Posse
Provincia: Bs.As
Ciudad: Capital Federal
------------------
Nombre: Jorge
Apellido: Guzmán
Provincia: Buenos Aires
Ciudad: Capital Federal
----------------------------
Nombre: Gonzalo 
Apellido: Posse
Provincia: Córdoba CBA Cba Cordoba Cordova
Ciudad: Capital Federal

FoozyWoozy
"""